### ISSUE TITLE
Docker multiple processes

### ISSUE THREAD

**Issue by olatarkowska** [common-workflow-language/common-workflow-language#391]
EDITED: I was wondering how CWL **scale up**. So my example would be how to split input data to batches and run the same analysis workflow in parallel (ideally on external resources). Is there any workflow manager? from the docker example I see that only run locally

In more complex pipeline I see a need for running multiple processes (e.g. horizontal sharing input data between workflow steps, etc). The only way to run multiple dummy docker containers seems to be using `docker-compose`:

_docker.cwl:_
```
cwlVersion: cwl:draft-3
class: CommandLineTool
baseCommand: [docker-compose]
inputs:
  - id: src
    type: File
    inputBinding:
      position: 1
      prefix: -f
  - id: dc_cmd
    type: string
    inputBinding:
      position: 2
  - id: dc_build
    type: boolean
    inputBinding:
      position: 3
      prefix: --build
  - id: dc_exit
    type: boolean
    inputBinding:
      position: 4
      prefix: --abort-on-container-exit
outputs: []
```

_docker-job.yml:_
```
src:
  class: File
  path: docker-compose.yml
dc_cmd:  up
dc_build: True
dc_exit: True
```

_docker-compose.yml_
```
version: '2'
services:
  test1:
    image: postgres
  test2:
    image: postgres
    command: /bin/bash -c "until psql -h test1 -U postgres -c '\l'; do sleep 5; done; echo connected"
```

**running:**
```
$ cwl-runner docker.cwl docker-job.yml 
/Users/ola/CWL/cwl-venv/bin/cwl-runner 1.0.20170217172322
Resolved 'docker.cwl' to 'file:///Users/ola/CWL/docker/docker.cwl'
[job docker.cwl] /private/var/folders/xl/jj4w78c127n_2z4ww0zyrvf80000gn/T/tmph2Nh2b$ docker-compose \
    -f \
    /private/var/folders/xl/jj4w78c127n_2z4ww0zyrvf80000gn/T/tmpHrP35x/stg258c802a-eccf-4f31-b0e8-9cdf0efd33fb/docker-compose.yml \
    up \
    --build \
    --abort-on-container-exit
Creating network "stg258c802aeccf4f31b0e89cdf0efd33fb_default" with the default driver
Creating stg258c802aeccf4f31b0e89cdf0efd33fb_test2_1
Creating stg258c802aeccf4f31b0e89cdf0efd33fb_test1_1
Attaching to stg258c802aeccf4f31b0e89cdf0efd33fb_test2_1, stg258c802aeccf4f31b0e89cdf0efd33fb_test1_1
test2_1  | psql: could not connect to server: Connection refused
test2_1  | 	Is the server running on host "test1" (192.168.192.2) and accepting
test2_1  | 	TCP/IP connections on port 5432?
test1_1  | The files belonging to this database system will be owned by user "postgres".
test1_1  | This user must also own the server process.
test1_1  | 
test1_1  | The database cluster will be initialized with locale "en_US.utf8".
test1_1  | The default database encoding has accordingly been set to "UTF8".
test1_1  | The default text search configuration will be set to "english".
test1_1  | 
test1_1  | Data page checksums are disabled.
test1_1  | 
test1_1  | fixing permissions on existing directory /var/lib/postgresql/data ... ok
test1_1  | creating subdirectories ... ok
test1_1  | selecting default max_connections ... 100
test1_1  | selecting default shared_buffers ... 128MB
test1_1  | selecting dynamic shared memory implementation ... posix
test1_1  | creating configuration files ... ok
test1_1  | running bootstrap script ... ok
test1_1  | performing post-bootstrap initialization ... ok
test1_1  | syncing data to disk ... ok
test1_1  | 
test1_1  | Success.
...
test1_1  | PostgreSQL init process complete; ready for start up.
test1_1  | 
test1_1  | LOG:  database system was shut down at 2017-02-28 05:38:47 UTC
test1_1  | LOG:  MultiXact member wraparound protections are now enabled
test1_1  | LOG:  database system is ready to accept connections
test1_1  | LOG:  autovacuum launcher started
test2_1  |                                  List of databases
test2_1  |    Name    |  Owner   | Encoding |  Collate   |   Ctype    |   Access privileges   
test2_1  | -----------+----------+----------+------------+------------+-----------------------
test2_1  |  postgres  | postgres | UTF8     | en_US.utf8 | en_US.utf8 | 
test2_1  |  template0 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
test2_1  |            |          |          |            |            | postgres=CTc/postgres
test2_1  |  template1 | postgres | UTF8     | en_US.utf8 | en_US.utf8 | =c/postgres          +
test2_1  |            |          |          |            |            | postgres=CTc/postgres
test2_1  | (3 rows)
test2_1  | 
test2_1  | connected
stg258c802aeccf4f31b0e89cdf0efd33fb_test2_1 exited with code 0
Aborting on container exit...
Stopping stg258c802aeccf4f31b0e89cdf0efd33fb_test1_1 ... done
[job docker.cwl] completed success
{}
Final process status is success
```

Are you planning to support something like that? How that could also scale up?

**Comment by StarvingMarvin** [comment]
Depending on what are you mining by 'something like that', it a) already exists, b) is explicitly not part of the spec or c) doesn't exist yet, but there are some vague plans to add it.

So... Is there a way to run multiple jobs in parallel: yes. The goal of the spec is to make possible to describe your apps/workflows in a way that makes possible for implementations to smartly schedule apps, without mandating any particular scheduler implementation. You can for example annotate that your tool is single threaded and uses 2GB of ram, and than make workflow that scatters that tool on each file/sample/whatever meaningful group. Implementation can that decide to run appropriate number of individual jobs in parallel. Now, reference implementation when being executed locally, probably just run things sequentially, but other implementations have ability to schedule multiple jobs on single instance or multiple instances as they see fit.

Is there a way to run a service, like postgresql in your example, that is alive for a while while other things depend on it: no. Right now tools are defined as "non-interactive command line tools". There is a rough consensus that such a thing should be part of the spec, but no one had a specific use case that would bring this discussion from the realm of theory. So if this is what you are asking, feel free to elaborate exact use case/semantics, and eventually demonstrate an implementation.

**Comment by mr-c** [comment]
Hello @olatarkowska We haven't heard from you in a while, so I'm going to close this issue for now. Please feel free to re-open!
